File: packages/shared-schemas/src/actions.ts
import { z } from "zod";

import { MetricFactSchema } from "./connectors";

export const ContentDraftRevisionBindingSchema = z
  .object({
    work_item_id: z.string().min(1),
    handoff_id: z.string().min(1),
    revision_id: z.string().min(1),
    content_digest: z.string().regex(/^[0-9a-f]{64}$/),
    draft_package_id: z.string().min(1),
    draft_package_digest: z.string().regex(/^[0-9a-f]{64}$/),
    approval_decision_id: z.string().min(1),
    final_canonical_url: z.string().min(1)
  })
  .strict();

export const ActionWordPressDraftApplyBlockerSchema = z.object({
  code: z.string().min(1),
  label: z.string().min(1),
  reason: z.string().min(1),
  next_step: z.string().min(1)
});

export const AuditEventSchema = z.object({
  id: z.string(),
  action_id: z.string().nullable().optional(),
  event_type: z.string(),
  event_type_label: z.string().default(""),
  actor: z.string(),
  created_at: z.string(),
  summary: z.string(),
  evidence_ids: z.array(z.string()),
  details: z.record(z.string(), z.unknown()).default({}),
  redacted: z.boolean()
});

export const ActionReviewOutcomeSchema = z.enum([
  "approved_for_prepare",
  "needs_changes",
  "rejected",
  "deferred"
]);

export const ActionReviewRequestSchema = z.object({
  outcome: ActionReviewOutcomeSchema,
  reviewed_by: z.string().min(1),
  notes: z.string().min(1).max(2000),
  checked_items: z.array(z.string()).default([]),
  blockers: z.array(z.string()).default([]),
  wordpress_draft: ContentDraftRevisionBindingSchema.nullable().optional()
});

export const ActionReviewGateSchema = z.object({
  status: z
    .enum(["pending_validation", "validated_prepare_only", "ready_to_apply", "blocked_apply"])
    .default("pending_validation"),
  status_label: z.string().default(""),
  summary: z.string().default("Wymaga sprawdzenia w WILQ przed kolejnym krokiem."),
  required_checks: z.array(z.string()).default([]),
  required_check_labels: z.array(z.string()).default([]),
  operator_checklist: z.array(z.string()).default([]),
  operator_checklist_labels: z.array(z.string()).default([]),
  apply_blockers: z.array(z.string()).default([]),
  apply_blocker_labels: z.array(z.string()).default([]),
  apply_blocker_summary_label: z.string().optional().default(""),
  confirmation_required: z.boolean().default(true),
  apply_allowed: z.boolean().default(false),
  last_review_outcome: ActionReviewOutcomeSchema.nullable().optional(),
  last_review_outcome_label: z.string().nullable().optional(),
  last_reviewed_by: z.string().nullable().optional(),
  last_reviewed_at: z.string().nullable().optional(),
  last_review_summary: z.string().nullable().optional(),
  last_confirmation_by: z.string().nullable().optional(),
  last_confirmation_at: z.string().nullable().optional(),
  last_confirmation_summary: z.string().nullable().optional(),
  last_impact_check_status: z.enum(["checked", "blocked"]).nullable().optional(),
  last_impact_check_status_label: z.string().nullable().optional(),
  last_impact_checked_by: z.string().nullable().optional(),
  last_impact_checked_at: z.string().nullable().optional(),
  last_impact_check_summary: z.string().nullable().optional(),
  last_mutation_audit_id: z.string().nullable().optional(),
  last_mutation_audit_status: z.enum(["blocked", "applied", "failed"]).nullable().optional(),
  last_mutation_audit_status_label: z.string().nullable().optional(),
  last_mutation_audit_actor: z.string().nullable().optional(),
  last_mutation_audit_at: z.string().nullable().optional(),
  last_mutation_audit_summary: z.string().nullable().optional(),
  last_mutation_adapter_reached: z.boolean().nullable().optional(),
  last_mutation_adapter_reached_label: z.string().nullable().optional(),
  last_external_write_attempted: z.boolean().nullable().optional(),
  last_external_write_attempted_label: z.string().nullable().optional(),
  last_mutation_attempted: z.boolean().nullable().optional(),
  last_mutation_attempted_label: z.string().nullable().optional(),
  last_mutation_adapter: z.string().nullable().optional(),
  last_mutation_adapter_label: z.string().nullable().optional(),
  last_mutation_audit_event_id: z.string().nullable().optional(),
  last_mutation_audit_trace_label: z.string().nullable().optional(),
  last_mutation_blockers: z.array(z.string()).default([]),
  last_mutation_blocker_labels: z.array(z.string()).default([]),
  last_mutation_blocker_summary_label: z.string().optional().default("")
});

const DEFAULT_ACTION_REVIEW_GATE = {
  status: "pending_validation",
  status_label: "",
  summary: "Wymaga sprawdzenia w WILQ przed kolejnym krokiem.",
  required_checks: [],
  required_check_labels: [],
  operator_checklist: [],
  operator_checklist_labels: [],
  apply_blockers: [],
  apply_blocker_labels: [],
  apply_blocker_summary_label: "",
  confirmation_required: true,
  apply_allowed: false,
  last_mutation_blockers: [],
  last_mutation_blocker_labels: [],
  last_mutation_blocker_summary_label: ""
} satisfies z.input<typeof ActionReviewGateSchema>;

export const ActionPreviewRowViewModelSchema = z.object({
  label: z.string(),
  value: z.string()
});

export const ActionPreviewCardViewModelSchema = z.object({
  id: z.string(),
  kind: z.string(),
  title_label: z.string(),
  subtitle_label: z.string().default(""),
  status_label: z.string().default(""),
  rows: z.array(ActionPreviewRowViewModelSchema).default([]),
  apply_state_label: z.string().default(""),
  system_readiness_label: z.string().default("")
});

export const ActionPreviewItemViewModelSchema = z.object({
  id: z.string(),
  title_label: z.string(),
  status_label: z.string().default(""),
  rows: z.array(ActionPreviewRowViewModelSchema).default([])
});

export const ActionDomainSchema = z.enum([
  "google_ads",
  "gsc_seo",
  "ahrefs",
  "localo",
  "wordpress",
  "social",
  "knowledge",
  "content",
  "codex",
  "ga4",
  "merchant",
  "google_sheets"
]);

export const ActionStatusSchema = z.enum([
  "new",
  "ready",
  "needs_validation",
  "validation_failed",
  "ready_to_apply",
  "applying",
  "applied",
  "failed",
  "dismissed",
  "blocked"
]);

export const ActionValidationStatusSchema = z.enum([
  "not_validated",
  "valid",
  "invalid"
]);

export const ActionObjectSchema = z.object({
  id: z.string(),
  title: z.string(),
  domain: ActionDomainSchema,
  connector: z.string(),
  connector_label: z.string().default(""),
  mode: z.enum(["suggest", "prepare", "apply"]),
  mode_label: z.string().default(""),
  risk: z.enum(["low", "medium", "high", "critical"]),
  risk_label: z.string().default(""),
  status: ActionStatusSchema,
  status_label: z.string().default(""),
  evidence_ids: z.array(z.string()).min(1),
  evidence_summary_label: z.string().default(""),
  metrics: z.array(MetricFactSchema),
  human_diagnosis: z.string(),
  recommended_reason: z.string(),
  validation_status: ActionValidationStatusSchema,
  validation_status_label: z.string().default(""),
  review_gate: ActionReviewGateSchema.optional().default(DEFAULT_ACTION_REVIEW_GATE),
  preview_cards: z.array(ActionPreviewCardViewModelSchema).default([]),
  payload: z.record(z.string(), z.unknown()),
  audit_events: z.array(AuditEventSchema)
});

export const ActionValidationResultSchema = z.object({
  action_id: z.string(),
  valid: z.boolean(),
  status: z.enum(["valid", "invalid"]),
  status_label: z.string().default(""),
  errors: z.array(z.string()),
  warnings: z.array(z.string()),
  checked_at: z.string()
});

export const ActionMutationAuditRecordSchema = z.object({
  id: z.string(),
  action_id: z.string(),
  connector: z.string(),
  action_type: z.string().nullable().optional(),
  status: z.enum(["blocked", "applied", "failed"]),
  status_label: z.string().default(""),
  adapter_reached: z.boolean().default(false),
  external_write_attempted: z.boolean().default(false),
  mutation_attempted: z.boolean(),
  mutation_adapter: z.string().nullable().optional(),
  actor: z.string(),
  created_at: z.string(),
  audit_event_id: z.string(),
  evidence_ids: z.array(z.string()),
  blockers: z.array(z.string()),
  wordpress_draft_binding: ContentDraftRevisionBindingSchema.nullable().optional(),
  wordpress_revision_blockers: z.array(ActionWordPressDraftApplyBlockerSchema).default([]),
  summary: z.string(),
  redacted: z.boolean()
});

export const ActionMutationReadinessRequirementSchema = z.object({
  code: z.string(),
  label: z.string(),
  satisfied: z.boolean().default(false),
  evidence: z.string().nullable().optional()
});

export const ActionMutationReadinessBlockerSchema = z.object({
  code: z.string(),
  label: z.string(),
  reason: z.string(),
  next_step: z.string()
});

export const ActionMutationApplyContractSchema = z.object({
  contract: z.literal("action_apply_contract_v1"),
  action_id: z.string(),
  action_type: z.string(),
  connector: z.string(),
  allowed_operation: z.string(),
  required_mode: z.literal("apply").default("apply"),
  draft_only: z.boolean().default(true),
  publication_allowed: z.boolean().default(false),
  destructive_allowed: z.boolean().default(false),
  adapter_status: z.enum(["not_implemented", "implemented"]).default("not_implemented"),
  required_env_flags: z.array(z.string()).default([]),
  required_input_contracts: z.array(z.string()).default([]),
  required_audit_events: z.array(z.string()).default([]),
  blocked_outputs: z.array(z.string()).default([]),
  operator_summary: z.string()
});

export const ActionMutationReadinessResponseSchema = z.object({
  response_type: z.literal("action_mutation_readiness"),
  contract: z.literal("action_mutation_readiness_v1"),
  action_id: z.string(),
  title: z.string(),
  connector: z.string(),
  connector_label: z.string().default(""),
  mode: z.enum(["suggest", "prepare", "apply"]),
  mode_label: z.string().default(""),
  risk: z.enum(["low", "medium", "high", "critical"]),
  risk_label: z.string().default(""),
  validation_status: ActionValidationStatusSchema,
  review_gate_status: z.string().default(""),
  ready_to_request_apply: z.boolean().default(false),
  vendor_write_possible: z.boolean().default(false),
  would_attempt_vendor_write: z.boolean().default(false),
  mutation_adapter: z.string().nullable().optional(),
  apply_contract: ActionMutationApplyContractSchema.nullable().optional(),
  target_candidate_id: z.string().nullable().optional(),
  target_label: z.string().nullable().optional(),
  target_url: z.string().nullable().optional(),
  write_authorization_status: z
    .enum([
      "missing_audit_trace",
      "audit_actor_mismatch",
      "available",
      "blocked_outside_action_apply"
    ])
    .nullable()
    .optional(),
  missing_audit_event_types: z.array(z.string()).default([]),
  requirements: z.array(ActionMutationReadinessRequirementSchema).default([]),
  blockers: z.array(ActionMutationReadinessBlockerSchema).default([]),
  operator_next_step: z.string(),
  evidence_ids: z.array(z.string()).default([]),
  source_connectors: z.array(z.string()).default([]),
  latest_mutation_audit_id: z.string().nullable().optional(),
  latest_mutation_audit_status: z.enum(["blocked", "applied", "failed"]).nullable().optional()
});

export const ActionMutationReadinessSummaryResponseSchema = z.object({
  response_type: z.literal("action_mutation_readiness_summary"),
  contract: z.literal("action_mutation_readiness_summary_v1"),
  action_count: z.number().int().nonnegative().default(0),
  ready_to_request_apply_count: z.number().int().nonnegative().default(0),
  vendor_write_possible_count: z.number().int().nonnegative().default(0),
  would_attempt_vendor_write_count: z.number().int().nonnegative().default(0),
  prepare_only_count: z.number().int().nonnegative().default(0),
  missing_adapter_count: z.number().int().nonnegative().default(0),
  high_risk_blocked_count: z.number().int().nonnegative().default(0),
  top_blockers: z.array(z.string()).default([]),
  first_write_candidate: ActionMutationReadinessResponseSchema.nullable().optional(),
  first_write_candidate_reason: z.string().default(""),
  activation_plan_steps: z.array(z.string()).default([]),
  activation_next_step: z.string().default(""),
  operator_next_step: z.string(),
  items: z.array(ActionMutationReadinessResponseSchema).default([])
});

export const ActionApplyResultSchema = z.object({
  action_id: z.string(),
  applied: z.boolean(),
  status: z.enum(["applied", "blocked", "failed"]),
  status_label: z.string().default(""),
  audit_event: AuditEventSchema,
  mutation_audit: ActionMutationAuditRecordSchema,
  errors: z.array(z.string()),
  wordpress_revision_blockers: z.array(ActionWordPressDraftApplyBlockerSchema).default([]),
  adapter_result: z.record(z.string(), z.unknown()).nullable().optional()
});

export const ActionWordPressDraftApplyInputSchema = ContentDraftRevisionBindingSchema;

export const ActionPreviewRequestSchema = z.object({
  requested_by: z.string().min(1).nullable().optional(),
  max_items: z.number().int().min(1).max(50).optional(),
  wordpress_draft: ContentDraftRevisionBindingSchema.nullable().optional()
});

export const ActionPreviewResultSchema = z.object({
  action_id: z.string(),
  status: z.enum(["preview_ready", "blocked"]),
  status_label: z.string().default(""),
  dry_run: z.boolean(),
  mutation_allowed: z.boolean(),
  preview_contract: z.string().nullable().optional(),
  preview_items: z.array(ActionPreviewItemViewModelSchema),
  preview_cards: z.array(ActionPreviewCardViewModelSchema).default([]),
  preview_items_total: z.number(),
  omitted_items: z.number(),
  blockers: z.array(z.string()),
  blocker_labels: z.array(z.string()).default([]),
  audit_event: AuditEventSchema,
  review_gate: ActionReviewGateSchema
});

export const ActionReviewResultSchema = z.object({
  action_id: z.string(),
  status: z.enum(["recorded"]),
  status_label: z.string().default(""),
  audit_event: AuditEventSchema,
  review_gate: ActionReviewGateSchema
});

export const ActionConfirmRequestSchema = z.object({
  confirmed_by: z.string().min(1),
  notes: z.string().min(1).max(2000),
  preview_acknowledged: z.boolean().default(false),
  target_roas: z.number().positive().nullable().optional(),
  target_cpa_micros: z.number().int().positive().nullable().optional(),
  wordpress_draft: ContentDraftRevisionBindingSchema.nullable().optional()
});

export const ActionConfirmResultSchema = z.object({
  action_id: z.string(),
  confirmed: z.boolean(),
  status: z.enum(["confirmed", "blocked"]),
  status_label: z.string().default(""),
  blockers: z.array(z.string()),
  blocker_labels: z.array(z.string()).default([]),
  audit_event: AuditEventSchema,
  review_gate: ActionReviewGateSchema
});

export const ActionImpactCheckRequestSchema = z.object({
  checked_by: z.string().min(1),
  notes: z.string().min(1).max(2000),
  pre_window_days: z.number().int().min(1).max(90).optional(),
  post_window_days: z.number().int().min(1).max(90).optional(),
  wordpress_draft: ContentDraftRevisionBindingSchema.nullable().optional()
});

export const ActionImpactCheckResultSchema = z.object({
  action_id: z.string(),
  status: z.enum(["checked", "blocked"]),
  status_label: z.string().default(""),
  pre_window_days: z.number(),
  post_window_days: z.number(),
  metric_fact_count: z.number(),
  source_connectors: z.array(z.string()),
  source_connector_labels: z.array(z.string()).default([]),
  evidence_ids: z.array(z.string()),
  evidence_summary_label: z.string().default(""),
  blockers: z.array(z.string()),
  blocker_labels: z.array(z.string()).default([]),
  audit_event: AuditEventSchema,
  review_gate: ActionReviewGateSchema
});

export const ActionApplyRequestSchema = z.object({
  confirm: z.boolean(),
  confirmed_by: z.string().min(1),
  wordpress_draft: ActionWordPressDraftApplyInputSchema.optional()
});

export type ActionPreviewCardViewModel = z.infer<typeof ActionPreviewCardViewModelSchema>;
export type ActionObject = z.infer<typeof ActionObjectSchema>;
export type ActionValidationResult = z.infer<typeof ActionValidationResultSchema>;
export type ActionMutationAuditRecord = z.infer<typeof ActionMutationAuditRecordSchema>;
export type ActionMutationReadinessResponse = z.infer<typeof ActionMutationReadinessResponseSchema>;
export type ActionMutationReadinessSummaryResponse = z.infer<
  typeof ActionMutationReadinessSummaryResponseSchema
>;
export type ActionApplyResult = z.infer<typeof ActionApplyResultSchema>;
export type ContentDraftRevisionBinding = z.infer<typeof ContentDraftRevisionBindingSchema>;
export type ActionWordPressDraftApplyBlocker = z.infer<
  typeof ActionWordPressDraftApplyBlockerSchema
>;
export type ActionWordPressDraftApplyInput = z.infer<typeof ActionWordPressDraftApplyInputSchema>;
export type ActionPreviewRequest = z.infer<typeof ActionPreviewRequestSchema>;
export type ActionPreviewResult = z.infer<typeof ActionPreviewResultSchema>;
export type ActionReviewRequest = z.infer<typeof ActionReviewRequestSchema>;
export type ActionReviewResult = z.infer<typeof ActionReviewResultSchema>;
export type ActionConfirmRequest = z.infer<typeof ActionConfirmRequestSchema>;
export type ActionConfirmResult = z.infer<typeof ActionConfirmResultSchema>;
export type ActionImpactCheckRequest = z.infer<typeof ActionImpactCheckRequestSchema>;
export type ActionImpactCheckResult = z.infer<typeof ActionImpactCheckResultSchema>;
export type ActionApplyRequest = z.infer<typeof ActionApplyRequestSchema>;

File: wilq/schemas/actions.py
from __future__ import annotations

from datetime import datetime
from typing import Any, Literal

from pydantic import BaseModel, ConfigDict, Field, field_validator, model_validator

from wilq.content.workflow.revision_binding import ContentDraftRevisionBinding
from wilq.operator_labels import blocker_count_label, impact_comparison_summary_label

from .core import ActionMode, ActionRisk, ActionStatus, MetricFact, OpportunityDomain, utc_now


class AuditEvent(BaseModel):
    id: str
    action_id: str | None = None
    event_type: str
    event_type_label: str = ""
    actor: str
    principal_id: str | None = None
    workspace_id: str | None = None
    trust_level: Literal["local_unverified"] | None = None
    submitted_actor_label: str | None = None
    created_at: datetime = Field(default_factory=utc_now)
    summary: str
    evidence_ids: list[str] = Field(default_factory=list)
    details: dict[str, Any] = Field(default_factory=dict)
    redacted: bool = True


class ActionWordPressDraftApplyBlocker(BaseModel):
    model_config = ConfigDict(extra="forbid")

    code: str = Field(min_length=1)
    label: str = Field(min_length=1)
    reason: str = Field(min_length=1)
    next_step: str = Field(min_length=1)


class ActionMutationAuditRecord(BaseModel):
    id: str
    action_id: str
    connector: str
    action_type: str | None = None
    status: Literal["blocked", "applied", "failed"]
    status_label: str = ""
    adapter_reached: bool = False
    external_write_attempted: bool = False
    mutation_attempted: bool = False
    mutation_adapter: str | None = None
    actor: str
    principal_id: str | None = None
    workspace_id: str | None = None
    trust_level: Literal["local_unverified"] | None = None
    submitted_actor_label: str | None = None
    created_at: datetime = Field(default_factory=utc_now)
    audit_event_id: str
    evidence_ids: list[str] = Field(default_factory=list)
    blockers: list[str] = Field(default_factory=list)
    wordpress_draft_binding: ContentDraftRevisionBinding | None = None
    wordpress_revision_blockers: list[ActionWordPressDraftApplyBlocker] = Field(
        default_factory=list
    )
    summary: str
    redacted: bool = True


class ActionMutationReadinessRequirement(BaseModel):
    code: str
    label: str
    satisfied: bool = False
    evidence: str | None = None


class ActionMutationReadinessBlocker(BaseModel):
    code: str
    label: str
    reason: str
    next_step: str


class ActionMutationApplyContract(BaseModel):
    contract: Literal["action_apply_contract_v1"] = "action_apply_contract_v1"
    action_id: str
    action_type: str
    connector: str
    allowed_operation: str
    required_mode: Literal["apply"] = "apply"
    draft_only: bool = True
    publication_allowed: bool = False
    destructive_allowed: bool = False
    adapter_status: Literal["not_implemented", "implemented"] = "not_implemented"
    required_env_flags: list[str] = Field(default_factory=list)
    required_input_contracts: list[str] = Field(default_factory=list)
    required_audit_events: list[str] = Field(default_factory=list)
    blocked_outputs: list[str] = Field(default_factory=list)
    operator_summary: str


class ActionMutationReadinessResponse(BaseModel):
    response_type: Literal["action_mutation_readiness"] = "action_mutation_readiness"
    contract: Literal["action_mutation_readiness_v1"] = "action_mutation_readiness_v1"
    action_id: str
    title: str
    connector: str
    connector_label: str = ""
    mode: ActionMode
    mode_label: str = ""
    risk: ActionRisk
    risk_label: str = ""
    validation_status: Literal["not_validated", "valid", "invalid"]
    review_gate_status: str = ""
    ready_to_request_apply: bool = False
    vendor_write_possible: bool = False
    would_attempt_vendor_write: bool = False
    mutation_adapter: str | None = None
    apply_contract: ActionMutationApplyContract | None = None
    target_candidate_id: str | None = None
    target_label: str | None = None
    target_url: str | None = None
    write_authorization_status: (
        Literal[
            "missing_audit_trace",
            "audit_actor_mismatch",
            "available",
            "blocked_outside_action_apply",
        ]
        | None
    ) = None
    missing_audit_event_types: list[str] = Field(default_factory=list)
    requirements: list[ActionMutationReadinessRequirement] = Field(default_factory=list)
    blockers: list[ActionMutationReadinessBlocker] = Field(default_factory=list)
    operator_next_step: str
    evidence_ids: list[str] = Field(default_factory=list)
    source_connectors: list[str] = Field(default_factory=list)
    latest_mutation_audit_id: str | None = None
    latest_mutation_audit_status: Literal["blocked", "applied", "failed"] | None = None


class ActionMutationReadinessSummaryResponse(BaseModel):
    response_type: Literal["action_mutation_readiness_summary"] = (
        "action_mutation_readiness_summary"
    )
    contract: Literal["action_mutation_readiness_summary_v1"] = (
        "action_mutation_readiness_summary_v1"
    )
    action_count: int = 0
    ready_to_request_apply_count: int = 0
    vendor_write_possible_count: int = 0
    would_attempt_vendor_write_count: int = 0
    prepare_only_count: int = 0
    missing_adapter_count: int = 0
    high_risk_blocked_count: int = 0
    top_blockers: list[str] = Field(default_factory=list)
    first_write_candidate: ActionMutationReadinessResponse | None = None
    first_write_candidate_reason: str = ""
    activation_plan_steps: list[str] = Field(default_factory=list)
    activation_next_step: str = ""
    operator_next_step: str
    items: list[ActionMutationReadinessResponse] = Field(default_factory=list)


ActionReviewOutcome = Literal[
    "approved_for_prepare",
    "needs_changes",
    "rejected",
    "deferred",
]


class ActionReviewRequest(BaseModel):
    outcome: ActionReviewOutcome
    reviewed_by: str = Field(min_length=1)
    notes: str = Field(min_length=1, max_length=2000)
    checked_items: list[str] = Field(default_factory=list)
    blockers: list[str] = Field(default_factory=list)
    wordpress_draft: ContentDraftRevisionBinding | None = None

    @field_validator("checked_items", "blockers")
    @classmethod
    def no_blank_review_items(cls, value: list[str]) -> list[str]:
        if any(not item.strip() for item in value):
            raise ValueError("review items must not be blank")
        return value


class ActionReviewGate(BaseModel):
    status: Literal[
        "pending_validation",
        "validated_prepare_only",
        "ready_to_apply",
        "blocked_apply",
    ] = "pending_validation"
    status_label: str = ""
    summary: str = "Wymaga walidacji akcji przed kolejnym krokiem."
    required_checks: list[str] = Field(default_factory=list)
    required_check_labels: list[str] = Field(default_factory=list)
    operator_checklist: list[str] = Field(default_factory=list)
    operator_checklist_labels: list[str] = Field(default_factory=list)
    apply_blockers: list[str] = Field(default_factory=list)
    apply_blocker_labels: list[str] = Field(default_factory=list)
    apply_blocker_summary_label: str = ""
    confirmation_required: bool = True
    apply_allowed: bool = False
    last_review_outcome: ActionReviewOutcome | None = None
    last_review_outcome_label: str | None = None
    last_reviewed_by: str | None = None
    last_reviewed_at: datetime | None = None
    last_review_summary: str | None = None
    last_confirmation_by: str | None = None
    last_confirmation_at: datetime | None = None
    last_confirmation_summary: str | None = None
    last_impact_check_status: Literal["checked", "blocked"] | None = None
    last_impact_check_status_label: str | None = None
    last_impact_checked_by: str | None = None
    last_impact_checked_at: datetime | None = None
    last_impact_check_summary: str | None = None
    last_mutation_audit_id: str | None = None
    last_mutation_audit_status: Literal["blocked", "applied", "failed"] | None = None
    last_mutation_audit_status_label: str | None = None
    last_mutation_audit_actor: str | None = None
    last_mutation_audit_at: datetime | None = None
    last_mutation_audit_summary: str | None = None
    last_mutation_adapter_reached: bool | None = None
    last_mutation_adapter_reached_label: str | None = None
    last_external_write_attempted: bool | None = None
    last_external_write_attempted_label: str | None = None
    last_mutation_attempted: bool | None = None
    last_mutation_attempted_label: str | None = None
    last_mutation_adapter: str | None = None
    last_mutation_adapter_label: str | None = None
    last_mutation_audit_event_id: str | None = None
    last_mutation_audit_trace_label: str | None = None
    last_mutation_blockers: list[str] = Field(default_factory=list)
    last_mutation_blocker_labels: list[str] = Field(default_factory=list)
    last_mutation_blocker_summary_label: str = ""

    @model_validator(mode="after")
    def hydrate_summary_labels(self) -> ActionReviewGate:
        if not self.apply_blocker_summary_label:
            self.apply_blocker_summary_label = blocker_count_label(
                self.apply_blocker_labels or self.apply_blockers
            )
        if not self.last_mutation_blocker_summary_label:
            self.last_mutation_blocker_summary_label = blocker_count_label(
                self.last_mutation_blocker_labels or self.last_mutation_blockers
            )
        self.last_impact_check_summary = impact_comparison_summary_label(
            self.last_impact_check_summary
        )
        return self


class ActionPreviewRowViewModel(BaseModel):
    label: str
    value: str


class ActionPreviewCardViewModel(BaseModel):
    id: str
    kind: str
    title_label: str
    subtitle_label: str = ""
    status_label: str = ""
    rows: list[ActionPreviewRowViewModel] = Field(default_factory=list)
    apply_state_label: str = ""
    system_readiness_label: str = ""


class ActionPreviewItemViewModel(BaseModel):
    id: str
    preview_contract: str | None = None
    candidate_id: str | None = None
    title_label: str
    status_label: str = ""
    rows: list[ActionPreviewRowViewModel] = Field(default_factory=list)


class ActionObject(BaseModel):
    model_config = ConfigDict(extra="forbid")

    id: str
    title: str
    domain: OpportunityDomain
    connector: str
    connector_label: str = ""
    mode: ActionMode
    mode_label: str = ""
    risk: ActionRisk
    risk_label: str = ""
    status: ActionStatus
    status_label: str = ""
    evidence_ids: list[str] = Field(min_length=1)
    evidence_summary_label: str = ""
    metrics: list[MetricFact] = Field(default_factory=list)
    human_diagnosis: str = Field(min_length=1)
    recommended_reason: str
    payload: dict[str, Any]
    validation_status: Literal["not_validated", "valid", "invalid"]
    validation_status_label: str = ""
    review_gate: ActionReviewGate = Field(default_factory=ActionReviewGate)
    preview_cards: list[ActionPreviewCardViewModel] = Field(default_factory=list)
    created_by: str
    created_at: datetime = Field(default_factory=utc_now)
    updated_at: datetime = Field(default_factory=utc_now)
    audit_events: list[AuditEvent] = Field(default_factory=list)

    @field_validator("evidence_ids")
    @classmethod
    def evidence_ids_not_blank(cls, value: list[str]) -> list[str]:
        if any(not item.strip() for item in value):
            raise ValueError("Identyfikatory dowodów akcji nie mogą być puste")
        return value


class ActionValidationResult(BaseModel):
    action_id: str
    valid: bool
    status: Literal["valid", "invalid"]
    status_label: str = ""
    errors: list[str] = Field(default_factory=list)
    warnings: list[str] = Field(default_factory=list)
    checked_at: datetime = Field(default_factory=utc_now)


class ActionApplyResult(BaseModel):
    action_id: str
    applied: bool
    status: Literal["applied", "blocked", "failed"]
    status_label: str = ""
    audit_event: AuditEvent
    mutation_audit: ActionMutationAuditRecord
    errors: list[str] = Field(default_factory=list)
    wordpress_revision_blockers: list[ActionWordPressDraftApplyBlocker] = Field(
        default_factory=list
    )
    adapter_result: dict[str, Any] | None = None


class ActionPreviewRequest(BaseModel):
    requested_by: str | None = None
    max_items: int = Field(default=8, ge=1, le=50)
    wordpress_draft: ContentDraftRevisionBinding | None = None


class ActionPreviewResult(BaseModel):
    action_id: str
    status: Literal["preview_ready", "blocked"]
    status_label: str = ""
    dry_run: bool = True
    mutation_allowed: bool = False
    preview_contract: str | None = None
    preview_items: list[ActionPreviewItemViewModel] = Field(default_factory=list)
    preview_cards: list[ActionPreviewCardViewModel] = Field(default_factory=list)
    preview_items_total: int = 0
    omitted_items: int = 0
    blockers: list[str] = Field(default_factory=list)
    blocker_labels: list[str] = Field(default_factory=list)
    audit_event: AuditEvent
    review_gate: ActionReviewGate


class ActionReviewResult(BaseModel):
    action_id: str
    status: Literal["recorded"]
    status_label: str = ""
    audit_event: AuditEvent
    review_gate: ActionReviewGate


class AdsExternalExecutionAcknowledgementRequest(BaseModel):
    """Human report that an Ads recommendation was executed outside WILQ."""

    model_config = ConfigDict(extra="forbid")

    measurement_plan_id: str = Field(min_length=1)
    execution_status: Literal["executed", "not_executed", "unknown"]
    acknowledged_by: str = Field(min_length=1, max_length=200)
    executed_at: datetime | None = None
    notes: str = Field(min_length=1, max_length=2000)


class ActionConfirmRequest(BaseModel):
    confirmed_by: str = Field(min_length=1)
    notes: str = Field(min_length=1, max_length=2000)
    preview_acknowledged: bool = False
    target_roas: float | None = Field(default=None, gt=0)
    target_cpa_micros: int | None = Field(default=None, gt=0)
    wordpress_draft: ContentDraftRevisionBinding | None = None


class ActionConfirmResult(BaseModel):
    action_id: str
    confirmed: bool
    status: Literal["confirmed", "blocked"]
    status_label: str = ""
    blockers: list[str] = Field(default_factory=list)
    blocker_labels: list[str] = Field(default_factory=list)
    audit_event: AuditEvent
    review_gate: ActionReviewGate


class AdsTargetGuardrailConfirmation(BaseModel):
    id: str
    connector_id: Literal["google_ads"] = "google_ads"
    action_id: str
    target_roas: float | None = Field(default=None, gt=0)
    target_cpa_micros: int | None = Field(default=None, gt=0)
    confirmed_by: str = Field(min_length=1)
    notes: str = Field(min_length=1, max_length=2000)
    audit_event_id: str
    evidence_ids: list[str] = Field(default_factory=list)
    created_at: datetime = Field(default_factory=utc_now)

    @model_validator(mode="after")
    def exactly_one_target(self) -> AdsTargetGuardrailConfirmation:
        target_count = int(self.target_roas is not None) + int(self.target_cpa_micros is not None)
        if target_count != 1:
            raise ValueError("exactly one Ads target guardrail must be confirmed")
        return self


class AdsStrategyReviewRecord(BaseModel):
    id: str
    connector_id: Literal["google_ads"] = "google_ads"
    action_id: str
    outcome: ActionReviewOutcome
    reviewed_by: str = Field(min_length=1)
    notes: str = Field(min_length=1, max_length=2000)
    checked_items: list[str] = Field(default_factory=list)
    blockers: list[str] = Field(default_factory=list)
    audit_event_id: str
    evidence_ids: list[str] = Field(default_factory=list)
    created_at: datetime = Field(default_factory=utc_now)


class ActionImpactCheckRequest(BaseModel):
    checked_by: str = Field(min_length=1)
    notes: str = Field(min_length=1, max_length=2000)
    pre_window_days: int = Field(default=7, ge=1, le=90)
    post_window_days: int = Field(default=7, ge=1, le=90)
    wordpress_draft: ContentDraftRevisionBinding | None = None


class ActionImpactCheckResult(BaseModel):
    action_id: str
    status: Literal["checked", "blocked"]
    status_label: str = ""
    pre_window_days: int
    post_window_days: int
    metric_fact_count: int = 0
    source_connectors: list[str] = Field(default_factory=list)
    source_connector_labels: list[str] = Field(default_factory=list)
    evidence_ids: list[str] = Field(default_factory=list)
    evidence_summary_label: str = ""
    blockers: list[str] = Field(default_factory=list)
    blocker_labels: list[str] = Field(default_factory=list)
    audit_event: AuditEvent
    review_gate: ActionReviewGate


ActionWordPressDraftApplyInput = ContentDraftRevisionBinding


class ActionApplyRequest(BaseModel):
    confirm: bool = False
    confirmed_by: str | None = None
    wordpress_draft: ActionWordPressDraftApplyInput | None = None


class CodexRun(BaseModel):
    id: str
    skill: str | None = None
    hook: str | None = None
    source: str | None = None
    status: Literal["started", "completed", "failed", "blocked"]
    used_endpoints: list[str] = Field(default_factory=list)
    evidence_ids: list[str] = Field(default_factory=list)
    action_ids: list[str] = Field(default_factory=list)
    started_at: datetime = Field(default_factory=utc_now)
    completed_at: datetime | None = None
    error: str | None = None
